sys/kern/kasan.c

```diff
@@ -217,10 +221,32 @@ static void call_ctors(void) {
   }
 }
 
+void kasan_grow(vaddr_t maxkvaddr) {
+  size_t size = maxkvaddr - KASAN_MD_SANITIZED_START - _kasan_sanitized_size;
+  assert(size % (SUPERPAGESIZE << 3) == 0);
+
+  size_t num_pde = (size >> KASAN_SHADOW_SCALE_SHIFT) / SUPERPAGESIZE;
+  vaddr_t va = KASAN_MD_SHADOW_START + _kasan_shadow_size;
+
+  for (size_t i = 0; i < num_pde; ++i) {
+    for (size_t j = 0; j < PT_ENTRIES; ++j) {
+      vm_page_t *pg = vm_page_alloc(1);
+      paddr_t pa = pg->paddr;
+      pmap_kenter(va, pa, VM_PROT_READ | VM_PROT_WRITE, 0);
+      va += PAGESIZE;
+      _kasan_shadow_size += PAGESIZE;
+    }
+  }
+
+  kasan_mark_valid(
+    (const void *)(KASAN_MD_SANITIZED_START + _kasan_sanitized_size), size);
+  _kasan_sanitized_size += size;
+}
+
```

Comment:
I don't like how this function looks for several reasons:
- The alignment requirement for `size` seems arbitrary (and pretty large). I don't see any particular reason why we would want to grow the sanitized space only in chunks of `SUPERPAGESIZE << 3`.
- The nested loop seems pointless at this point, and only makes the code harder to read.

I tried to write a simpler version and this is what I came up with:
```c
static inline vaddr_t kasan_va_to_shad_va(vaddr_t va) {
  return KASAN_MD_SHADOW_START +
         (va - KASAN_MD_SANITIZED_START) / KASAN_SHADOW_SCALE_SIZE;
}

void kasan_grow(vaddr_t maxkvaddr) {
  assert(mtx_owned(&maxkvaddr_lock))
  maxkvaddr = roundup2(maxkvaddr, PAGESIZE * KASAN_SHADOW_SCALE_SIZE);
  vaddr_t va = kasan_va_to_shad_va(_kasan_sanitized_end);
  vaddr_t end = kasan_va_to_shad_va(maxkvaddr);

  /* Allocate and map shadow pages to cover the new KVA space. */
  for (; va < end; va += PAGESIZE) {
    vm_page_t *pg = vm_page_alloc(1);
    pmap_kenter(va, pg->paddr, VM_PROT_READ | VM_PROT_WRITE, 0);
  }

  if (maxkvaddr > _kasan_sanitized_end) {
    kasan_mark_valid(
      (const void *)(_kasan_sanitized_end), maxkvaddr - _kasan_sanitized_end);
    _kasan_sanitized_end = maxkvaddr;
  }
}
```

Note that this assumes that we replace `_kasan_sanitized_size` and `_kasan_shadow_size` with `_kasan_sanitized_end`.

Reply:
You are right. We decided with to do that for simplicity but now I see that removing that assumptions make code much cleaner.
I have applied your suggestion.
Aggressive rounding has been removed from `mips/pmap.c`.